core/src/main/java/org/elasticsearch/index/query/AndQueryBuilder.java

```diff
@@ -111,7 +111,7 @@ public QueryValidationException validate() {
     }
 
     @Override
-    public String getName() {
+    public String getWriteableName() {
```

Comment:
For the agg refactoring I am doing the following: https://github.com/colings86/elasticsearch/blob/enhancement/minAggRefactor/core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java#L306 basically the super class defines a template method which is implemented by the sub-class to create the instance and then the super-class can call the methods it needs on that object to set what it needs to.